### PR TITLE
Move `PartialEq` impls behind DCOU feature

### PR DESCRIPTION
**Pull request by jstarry**
#### Problem
Implementing `PartialEq` for epoch stakes is quite expensive but it's not clear that it's actually okay-ish for it to be that way since equality is only ever checked in tests. For context, this issue came up here: https://github.com/anza-xyz/agave/pull/1146/files#r1594226939

#### Summary of Changes
Move some `PartialEq` impl's behind the `"dev-context-only-utils"` feature

Fixes #

### CODE REVIEW

File: runtime/src/stakes.rs [hunk 1]
```diff
@@ -491,6 +491,15 @@ impl StakesEnum {
...
 }
 
+impl From<StakesEnum> for Stakes<Delegation> {
+    fn from(stakes: StakesEnum) -> Self {
+        match stakes {
+            StakesEnum::Accounts(stakes) => stakes.into(),
+            StakesEnum::Delegations(stakes) => stakes,
+        }
+    }
+}
```
Comment (brooksprumo) [inline runtime/src/stakes.rs]: I am fine with this impl, but does seem orthogonal to solving this problem described by the PR. IMO we keep this PR focused to only `PartialEq` and DCOU.

File: runtime/src/stakes.rs [hunk 2]
```diff
@@ -525,19 +534,17 @@ impl From<Stakes<Delegation>> for StakesEnum {
...
-            (Self::Accounts(stakes), Self::Delegations(other)) => {
+            (Self::Delegations(stakes), Self::Delegations(other)) => stakes == other,
+            (stakes, other) => {
                 let stakes = Stakes::<Delegation>::from(stakes.clone());
-                &stakes == other
-            }
-            (Self::Delegations(stakes), Self::Accounts(other)) => {
                 let other = Stakes::<Delegation>::from(other.clone());
-                stakes == &other
+                stakes == other
```
Comment (brooksprumo) [inline runtime/src/stakes.rs]: I believe this change is from #1146. Can it be reverted for this PR? That'll keep this PR cleaner, and only pertain to `PartialEq` and DCOU.

File: runtime/src/stakes.rs [hunk 3]
```diff
@@ -204,7 +204,7 @@ pub struct Stakes<T: Clone> {
 // Below type allows EpochStakes to include either a Stakes<StakeAccount> or
 // Stakes<Delegation> and so bypass the conversion cost between the two at the
 // epoch boundary.
-#[derive(Debug, AbiExample)]
+#[derive(Debug, AbiExample, Clone)]
```
Comment (brooksprumo) [inline runtime/src/stakes.rs]: I think we can remove the `Clone` derive in this PR once the PartialEq changes on StakesEnum are reverted.

(And when/if it needs to be re-added, we could put Clone behind DCOU too.)

File: runtime/src/stakes.rs (hunk 1 above)
Comment (jstarry) [reply runtime/src/stakes.rs]: Oh, it's actually the main thing I wanted this PR to solve and wrote that in the problem description. Would it help if I change the PR title to be more epoch stakes specific? Or do you just want me to split into two PR's?

Comment (jstarry) [reply runtime/src/stakes.rs]: Well, you're right that I really didn't mention changing the `PartialEq` impl at all. Sure, I'll move these changes into a follow-up PR

Comment (jstarry) [reply runtime/src/stakes.rs]: Ok, those changes are all removed now!

**Review by brooksprumo — APPROVED** [review]
:shipit: 

Thanks! I love a PR that's easy to say yes to!
